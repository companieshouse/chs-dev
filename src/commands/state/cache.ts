import { Args, Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Cache extends Command {

    static description = "Cache the state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { args, flags } = await this.parse(Cache);
        const timestamp: string = new Date(Date.now()).toISOString();
        const cacheName: string = args.name || timestamp;

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`save chs-dev state?`);
        };

        if (await handlePrompt()) {
            //
        }
    }

}
