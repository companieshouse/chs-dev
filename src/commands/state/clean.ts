import { Command, Config, Flags } from "@oclif/core";
import { StateManager } from "../../state/state-manager.js";
import ChsDevConfig from "../../model/Config.js";
import loadConfig from "../../helpers/config-loader.js";
import { DockerCompose } from "../../run/docker-compose.js";
import { confirm } from "../../helpers/user-input.js";

export default class Clean extends Command {

    static description = "Wipes the state of chs-dev and revert to a default state";

    static flags = {
        purge: Flags.boolean({
            char: "p",
            description: "Wipe state, volumes and images",
            default: false
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        const logger = {
            log: (msg: string) => this.log(msg)
        };
        this.chsDevConfig = loadConfig();
        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);

    }

    async run (): Promise<any> {
        const { flags: { purge } } = await this.parse(Clean);

        const handlePrompt = async (): Promise<boolean> => {
            return await confirm(`This will reset your chs-dev state and delete all unused docker volumes${purge ? " including images." : "."} Do you want to proceed?`);
        };

        if (await handlePrompt()) {
            const containerStatus = this.dockerCompose.getServiceStatuses();
            if (containerStatus !== undefined) {
                this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
            }
            this.stateManager.cleanState();

            await this.config.runHook("generate-runnable-docker-compose", {});

            this.dockerCompose.prune("volume");
            if (purge) {
                this.dockerCompose.prune("image");
            }

        }
    }

}
