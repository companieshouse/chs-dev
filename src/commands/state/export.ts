import { Args, Command, Config } from "@oclif/core";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { StateManager } from "../../state/state-manager.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};
const EXPORT_STATE_DIR = ".exported_state_cache";

export default class Export extends Command {
    static description = "Export current state of chs-dev into a saved file";

    static args = {
        name: Args.string({
            required: true,
            description: "Name of the exported file"
        })
    };

    private readonly stateManager: StateManager;
    private readonly chsDevConfig: ChsDevConfig;
    private readonly exportedStateDir: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        this.exportedStateDir = join(this.chsDevConfig.projectPath, EXPORT_STATE_DIR);

        if (!existsSync(this.exportedStateDir)) {
            mkdirSync(this.exportedStateDir, { recursive: true });
        }

        this.stateManager = new StateManager(this.chsDevConfig.projectPath);
    }

    async run (): Promise<void> {
        const { args: { name } } = await this.parse(Export);

        if (await confirm(`Export current state of chs-dev to file named: '${name}'?`)) {
            this.exportState(name);
        }

    }

    private exportState (exportFileName: string): void {
        const exportedFilenamePath = join(this.exportedStateDir, `${exportFileName}.yaml`);
        writeContentToFile(this.getExportData(), exportedFilenamePath);

        this.log(`Exported state destination: ${exportedFilenamePath}`);
    }

    private getExportData (): StateCache {
        const stateSnapshot = this.stateManager.snapshot;
        const dockerComposeSnapshot = getGeneratedDockerComposeFile(this.chsDevConfig.projectPath);
        return {
            state: {
                hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                snapshot: stateSnapshot
            },
            dockerCompose: {
                hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                snapshot: dockerComposeSnapshot
            }
        };

    }

}
