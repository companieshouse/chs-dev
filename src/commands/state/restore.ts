import { Args, Command, Config, Flags } from "@oclif/core";
import { existsSync } from "fs";
import { basename, join } from "path";
import yaml from "yaml";
import loadConfig from "../../helpers/config-loader.js";
import { readFileContent, writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { confirm } from "../../helpers/user-input.js";
import ChsDevConfig from "../../model/Config.js";
import { DockerCompose } from "../../run/docker-compose.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string
    }
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    }
};

export default class Restore extends Command {
    static description = "Restore and update the state files from saved cache or imported cache.";

    static args = {
        name: Args.string({
            required: false,
            description: "Name of the cache"
        })
    };

    static flags = {
        importCacheFrom: Flags.string({
            char: "i",
            description: "Path to the exported cache"
        })
    };

    private readonly chsDevConfig: ChsDevConfig;
    private readonly dockerCompose: DockerCompose;
    private readonly stateCacheFile: string;

    constructor (argv: string[], config: Config) {
        super(argv, config);

        this.chsDevConfig = loadConfig();
        const cacheDir = config.cacheDir;
        this.stateCacheFile = join(cacheDir, `${basename(this.chsDevConfig.projectName)}.state.yaml`);

        const logger = { log: (msg: string) => this.log(msg) };

        this.dockerCompose = new DockerCompose(this.chsDevConfig, logger);
    }

    async run (): Promise<void> {
        const { args, flags: { importCacheFrom } } = await this.parse(Restore);

        const containerStatus = this.dockerCompose.getServiceStatuses();
        if (containerStatus !== undefined) {
            this.error("Ensure all containers are stopped. Run: '$ chs-dev down'");
        }

        const cacheName = importCacheFrom ? "import-cache" : args.name;

        if (!cacheName || typeof cacheName === "undefined") {
            this.error("Please provide a valid cache name or use the --importCacheFrom or -i flag to restore from an imported cache.");
        }
        console.log(`Restoring state from cache: ${cacheName}`);
        if (importCacheFrom) {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromImport(importCacheFrom);
            }
        } else {
            if (await this.handlePrompt(cacheName)) {
                this.restoreFromSavedCache(cacheName);
            }
        }

    }

    private restoreFromSavedCache (cacheName: string): void {
        const cacheData = this.loadSavedCacheData();
        const namedCacheData = cacheData[cacheName];
        if (!namedCacheData) {
            this.error(`Cache with name '${cacheName}' not found.`);
        }
        const data = this.verifyCacheAuthencity(namedCacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from saved cache '${cacheName}'`);
    }

    private restoreFromImport (importCachePath: string): void {
        if (!existsSync(importCachePath)) {
            this.error(`Import cache file does not exist in location: ${importCachePath}`);
        }

        const cacheData = readFileContent(importCachePath) as StateCache;

        if (!cacheData || typeof cacheData !== "object") {
            this.error(`Invalid cache data in imported file: ${importCachePath}`);
        }

        const data = this.verifyCacheAuthencity(cacheData, (stateCache) => stateCache);

        this.restoreState(data);
        this.log(`Restored state from imported cache.`);
    }

    private verifyCacheAuthencity<T> (cacheData: StateCache, cacheSupplier: (stateCache: StateCache) => StateCache): StateCache {

        if (cacheData.state.hash === hashAlgorithm(yaml.stringify(cacheData.state.snapshot)) &&
            cacheData.dockerCompose.hash === hashAlgorithm(yaml.stringify(cacheData.dockerCompose.snapshot))) {
            return cacheSupplier(cacheData);
        }
        this.error("Cache data has been corrupted or touched.");
    }

    private restoreState ({ state, dockerCompose }: StateCache): void {
        // Restore state file
        const stateFilePath = join(this.chsDevConfig.projectPath, `.chs-dev.yaml`);
        writeContentToFile(state.snapshot, stateFilePath);

        // Restore docker-compose file
        const dockerComposeFilePath = join(this.chsDevConfig.projectPath, "docker-compose.yaml");
        writeContentToFile(dockerCompose.snapshot, dockerComposeFilePath);
    }

    private loadSavedCacheData (): Record<string, StateCache> {
        return readFileContent(this.stateCacheFile) as Record<string, StateCache> || {};
    }

    private async handlePrompt (cacheName: string): Promise<boolean> {
        const message = cacheName === "import-cache"
            ? `Restore from an imported cache?`
            : `Restore from a saved cache '${cacheName}'?`;
        return confirm(message);
    }
}
