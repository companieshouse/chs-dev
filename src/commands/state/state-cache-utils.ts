import { existsSync } from "fs";
import { join } from "path";
import yaml from "yaml";
import { readFileContent, writeContentToFile } from "../../helpers/file-utils.js";
import { hashAlgorithm } from "../../helpers/index.js";
import ChsDevConfig from "../../model/Config.js";
import { confirm } from "../../helpers/user-input.js";

type StateCache = {
    state: {
        snapshot: Record<string, any>;
        hash: string;
    };
    dockerCompose: {
        snapshot: Record<string, any>;
        hash: string;
    };
};
type CacheActions = "add" | "remove" | "wipe" | "import-file-cache" | "import-saved-cache";

export const EXPORT_STATE_DIR = "exported_state_cache";

export const handlePrompt = async (
    action:CacheActions,
    cacheName: string = "default-cache-name"
): Promise<boolean> => {
    const messages: Record<string, string> = {
        add: `This will save the cache as '${cacheName}' or overwrite if name already exists. Proceed?`,
        remove: `Do you want to delete the cache '${cacheName}'?`,
        wipe: "Do you want to delete all saved caches?",
        "import-file-cache": "Restore from an imported file cache?",
        "import-saved-cache": `Restore from a saved cache '${cacheName}'?`
    };
    return await confirm(messages[action]);
};

export const validateNameFormat = (name: string): void => {
    const regex = /^[a-zA-Z0-9_-]+$/;
    if (!regex.test(name)) {
        throw new Error(`Invalid cache name format: '${name}'. Only alphanumeric characters with underscores or hypens are allowed.`);
    }
};

export const validateCacheNameExists = (cacheData: Record<string, any>, cacheName: string): void => {
    if (!cacheData[cacheName]) {
        throw new Error(`Cache named ${cacheName} does not exist.`);
    }
};

export function verifyCacheAuthenticity (cacheData: StateCache): StateCache {
    if (
        cacheData.state.hash === hashAlgorithm(yaml.stringify(cacheData.state.snapshot)) &&
            cacheData.dockerCompose.hash === hashAlgorithm(yaml.stringify(cacheData.dockerCompose.snapshot))
    ) {
        return cacheData;
    }
    throw new Error("Cache data has been corrupted or touched.");
}

export function loadImportCache (importCachePath: string): StateCache {
    if (!existsSync(importCachePath)) {
        throw new Error(`Import cache file does not exist in location: ${importCachePath}`);
    }
    const cacheData = readFileContent(importCachePath) as StateCache;
    if (!cacheData || typeof cacheData !== "object") {
        throw new Error(`Invalid cache data in imported file: ${importCachePath}`);
    }
    return cacheData;
}

export function restoreStateFiles (config: ChsDevConfig, state: StateCache["state"], dockerCompose: StateCache["dockerCompose"]): void {
    const stateFilePath = join(config.projectPath, ".chs-dev.yaml");
    writeContentToFile(state.snapshot, stateFilePath);
    const dockerComposeFilePath = join(config.projectPath, "docker-compose.yaml");
    writeContentToFile(dockerCompose.snapshot, dockerComposeFilePath);
}
