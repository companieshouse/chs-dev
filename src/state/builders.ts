import { Dirent, readdirSync, readFileSync } from "fs";
import { join } from "path";
import BuilderDockerComposeSpec from "../model/BuilderDockerComposeSpec.js";
import yaml from "yaml";

/**
 * Loads the requested version of the builder specification from the inventory
 * @param projectPath path to the project the tool is being run out of
 * @param builderName name of the builder being requested
 * @param version version of the builder being used
 * @returns BuilderDockerComposeSpec representing the builder
 */
export const getBuilder = (projectPath: string, builderName: string | undefined, version: string | undefined): BuilderDockerComposeSpec => {
    if (typeof builders === "undefined") {
        builders = {
            repository: {
                v1: {
                    name: "repository",
                    version: "v1",
                    builderSpec: repositoryBuilderSpec
                }
            }
        };

        populateBuildersCache(projectPath, builders);
    }

    if (typeof builderName === "undefined" || builderName.length === 0) {
        return getBuilder(projectPath, "repository", undefined);
    }

    if (!(builderName in builders)) {
        throw new Error(`No builder found for ${builderName} builder`);
    }

    const builderVersions = builders[builderName];

    const builderVersion = typeof version === "undefined"
        ? builderVersions[findLatestVersion(builderVersions)]
        : builderVersions[version];

    if (typeof builderVersion === "undefined") {
        throw new Error(`No version ${version} of ${builderName} builder`);
    }

    return builderVersion;
};

export const getBuilders = (projectPath: string) => {
    if (typeof builders === "undefined") {
        builders = {
            repository: {
                v1: {
                    name: "repository",
                    version: "v1",
                    builderSpec: repositoryBuilderSpec
                }
            }
        };

        populateBuildersCache(projectPath, builders);
    }

    return {
        ...builders
    };
};

export const clearBuilders = () => {
    builders = undefined;
};

const repositoryBuilderSpec = yaml.stringify({
    services: {
        "<service>": {
            pull_policy: "build",
            build: {
                context: "<absolute_repository_path>",
                args: {
                    // eslint-disable-next-line no-template-curly-in-string
                    SSH_PRIVATE_KEY: "${SSH_PRIVATE_KEY}",
                    // eslint-disable-next-line no-template-curly-in-string
                    SSH_PRIVATE_KEY_PASSPHRASE: "${SSH_PRIVATE_KEY_PASSPHRASE}"
                }
            },
            develop: {
                watch: [
                    {
                        path: ".touch",
                        action: "rebuild"
                    }
                ]
            }
        }
    }
});

let builders: undefined | Record<string, Record<string, BuilderDockerComposeSpec>>;

const isBuilderFile = (potentialBuilderFile: Dirent) => {
    return potentialBuilderFile.isFile() && (
        potentialBuilderFile.name === "builder.docker-compose.yaml" ||
        potentialBuilderFile.name === "builder.docker-compose.yml"
    );
};
const findLatestVersion = (builderVersions: Record<string, BuilderDockerComposeSpec>) => {
    return Object.keys(builderVersions).sort().reverse()[0];
};

const populateBuildersCache = (projectPath: string, builders: Record<string, Record<string, BuilderDockerComposeSpec>>) => {
    const buildersDir = join(projectPath, "local/builders");
    const buildersDirContents: Dirent[] = readdirSync(
        buildersDir, {
            recursive: true,
            withFileTypes: true
        }
    );

    const builderFiles = buildersDirContents.filter(isBuilderFile);

    for (const builder of builderFiles) {
        const pathWithoutBase = builder.parentPath.replace(
            `${buildersDir}/`, ""
        );

        const pathParts = pathWithoutBase.split("/");

        const spec = readFileSync(join(builder.parentPath, builder.name));

        if (typeof builders[pathParts[0]] === "undefined") {
            builders[pathParts[0]] = {};
        }

        builders[pathParts[0]][pathParts[1]] = {
            builderSpec: spec.toString(),
            name: pathParts[0],
            version: pathParts[1]
        };
    }
};
