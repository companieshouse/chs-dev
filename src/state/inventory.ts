import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "fs";
import { basename, join } from "path";

import yaml from "yaml";
import glob from "glob";

import { createHash } from "crypto";
import { Service } from "../model/Service.js";
import { Module } from "../model/Module.js";
import { readServices } from "./service-reader.js";
import { DependencyNameResolver } from "./dependency-name-resolver.js";

interface InventoryCache {
  hash: string;
  services: Service[];
  modules: Module[]
}

export class Inventory {
    readonly inventoryCacheFile: string;
    private __serviceFiles: string[] | undefined;

    constructor (private path: string, cacheDir: string) {
        this.path = path;
        this.inventoryCacheFile = join(cacheDir, `${basename(path)}.inventory.yaml`);

        if (!existsSync(cacheDir)) {
            mkdirSync(cacheDir, {
                recursive: true
            });
        }
    }

    get modules (): Module[] {
        return this.getFromCache(inventoryCache => inventoryCache.modules);
    }

    get services (): Service[] {
        return this.getFromCache(inventoryCache => inventoryCache.services);
    }

    private getFromCache<T> (cacheSupplier: (inventoryCache: InventoryCache) => T): T {
        const inventoryCache = this.inventoryCache;

        if (typeof inventoryCache !== "undefined") {
            if (inventoryCache.hash === this.hashServiceFiles()) {
                return cacheSupplier(inventoryCache);
            }
        }

        this.updateCache();

        return this.getFromCache(cacheSupplier);
    }

    private updateCache () {
        const inventory = {
            hash: this.hashServiceFiles(),
            services: this.loadServices(),
            modules: this.loadModules()
        };

        writeFileSync(this.inventoryCacheFile, yaml.stringify(inventory));
    }

    private loadModules (): Module[] {
        return readdirSync(join(this.path, "services", "modules"), { withFileTypes: true })
            .filter(item => item.isDirectory())
            .map(item => {
                return {
                    name: item.name
                };
            });
    }

    private loadServices (): Service[] {
        const partialServices: Partial<Service>[] = this.serviceFiles.flatMap(readServices);

        const dependencyNameResolver = new DependencyNameResolver(partialServices);

        return partialServices.map(service => ({
            ...service,
            dependsOn: dependencyNameResolver.fullDependencyListIncludingTransitive(service.dependsOn as string[])
        } as Service));
    }

    private hashServiceFiles () {
        const sha256Hash = createHash("sha256");

        this.serviceFiles
            .map(serviceFile => readFileSync(serviceFile))
            .forEach((serviceFile:any) => {
                sha256Hash.update(serviceFile);
            });

        return sha256Hash.digest("hex");
    }

    private get inventoryCache (): InventoryCache | undefined {
        if (existsSync(this.inventoryCacheFile)) {
            return yaml.parse(readFileSync(this.inventoryCacheFile).toString("utf-8"));
        }

        return undefined;
    }

    private get serviceFiles (): string[] {
        if (typeof this.__serviceFiles === "undefined") {
            this.__serviceFiles = glob.sync(`${this.path}/services/**/*.docker-compose.yaml`);
        }
        return this.__serviceFiles;
    }
}
