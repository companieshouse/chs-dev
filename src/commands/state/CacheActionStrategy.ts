import yaml from "yaml";

import { getGeneratedDockerComposeFile } from "../../helpers/docker-compose-file.js";
import { hashAlgorithm } from "../../helpers/index.js";
import { existsSync, mkdirSync } from "fs";
import { join } from "path";
import { writeContentToFile } from "../../helpers/file-utils.js";
import { handlePrompt, validateCacheNameExists } from "./state-cache-utils.js";
import { StateManager } from "../../state/state-manager.js";

const EXPORT_STATE_DIR = ".exported_state_cache";

export interface AvailableCacheContext {
    logger: (message: string) => void;
  }

export interface WipeCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface RemoveCacheContext {
    stateCacheFile: string;
    logger: (message: string) => void;
}

export interface ExportCacheContext {
    projectPath: string;
    logger: (message: string) => void;
}

export interface AddCacheContext {
    stateManager: StateManager;
    stateCacheFile: string;
    projectPath: string;
    logger: (message: string) => void;
}
interface CacheActionStrategy<C> {
    execute(cache: C, cacheData: Record<string, any>, cacheName?: string): Promise<void> | void;
}

const handleExportCache = (context: ExportCacheContext, cacheData: Record<string, any>, cacheName: string): void => {
    const exportDir = join(context.projectPath, EXPORT_STATE_DIR);
    if (!existsSync(exportDir)) {
        mkdirSync(exportDir, { recursive: true });
    }

    const exportedFilenamePath = join(exportDir, `${cacheName}.yaml`);

    writeContentToFile(cacheData[cacheName], exportedFilenamePath);
    context.logger(`Exported cache ${cacheName} destination: '${exportedFilenamePath}'`);
};

export class AvailableCacheStrategy implements CacheActionStrategy<AvailableCacheContext> {
    execute (context: AvailableCacheContext, cacheData: Record<string, any>): void {
        const cacheNames = Object.keys(cacheData);
        if (cacheNames.length > 0) {
            context.logger("Available caches:");
            cacheNames.forEach((name: string) => context.logger(`- ${name}`));
        } else {
            context.logger("No cache available.");
        }
    }
}

export class WipeCacheStrategy implements CacheActionStrategy<WipeCacheContext> {
    async execute (context: WipeCacheContext, _cacheData: Record<string, any>): Promise<void> {
        const confirmed = await handlePrompt("wipe");
        if (confirmed) {
            writeContentToFile({}, context.stateCacheFile);
            context.logger("Wiped all caches");
        }
    }
}

export class RemoveCacheStrategy implements CacheActionStrategy<RemoveCacheContext> {
    async execute (context: RemoveCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);

        const confirmed = await handlePrompt("remove", _cacheName);
        if (confirmed) {
            delete _cacheData[_cacheName];
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Removed cache '${_cacheName}'`);

        }
    }
}

export class ExportCacheStrategy implements CacheActionStrategy<ExportCacheContext> {
    async execute (context: ExportCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        validateCacheNameExists(_cacheData, _cacheName);
        handleExportCache(context, _cacheData, _cacheName);

    }
}

export class AddCacheStrategy implements CacheActionStrategy<AddCacheContext> {
    async execute (context: AddCacheContext, _cacheData: Record<string, any>, _cacheName: string): Promise<void> {
        const confirmed = await handlePrompt("add", _cacheName);
        if (confirmed) {
            const stateSnapshot = context.stateManager.snapshot;
            const dockerComposeSnapshot = getGeneratedDockerComposeFile(context.projectPath);
            _cacheData[_cacheName] = {
                state: {
                    hash: hashAlgorithm(yaml.stringify(stateSnapshot)),
                    snapshot: stateSnapshot
                },
                dockerCompose: {
                    hash: hashAlgorithm(yaml.stringify(dockerComposeSnapshot)),
                    snapshot: dockerComposeSnapshot
                }
            };
            writeContentToFile(_cacheData, context.stateCacheFile);
            context.logger(`Saved cache as '${_cacheName}'`);
        }
    }
}
