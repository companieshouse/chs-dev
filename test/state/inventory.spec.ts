import { expect, jest } from "@jest/globals";
import { copySync } from "fs-extra";
import { join } from "path";
import fs from "fs";
import { Inventory, Service } from "../../src/state/inventory";
import { parse, stringify } from "yaml";

describe("Inventory", () => {
    let tempDir: string;
    let inventoryDir: string;
    let confDir: string;
    let inventoryConfFile: string;

    beforeEach(() => {
        tempDir = fs.mkdtempSync("inventory");
        inventoryDir = join(tempDir, "inventory");
        confDir = join(tempDir, "conf");
        inventoryConfFile = join(confDir, "inventory.yaml");

        copySync(
            join(
                process.cwd(), "test/data/inventory"
            ),
            inventoryDir
        );
    });

    afterEach(() => {
        // @ts-expect-error does exist
        fs.rmSync(tempDir, { recursive: true, force: true });
    });

    const addNewService = (moduleName: string, serviceName: string) => {
        const newModulePath = join(inventoryDir, "services/modules", moduleName);

        fs.mkdirSync(newModulePath);

        const newService = {
            services: {
                [serviceName]: {
                    labels: [
                        "- chs.repository.url=git@github.com:companieshouse/service-five.git"
                    ],
                    depends_on: {
                        mongo: {
                            condition: "service_healthy"
                        }
                    }
                }
            }
        };

        fs.writeFileSync(
            join(newModulePath, `${serviceName}.docker-compose.yaml`),
            stringify(newService)
        );
    };

    describe("constructor", () => {
        afterEach(() => {
            // @ts-expect-error
            fs.rmSync(confDir, { recursive: true, force: true });
        });

        it("creates conf directory if not in existence", () => {
            // eslint-disable-next-line no-new
            new Inventory(inventoryDir, confDir);

            expect(fs.existsSync(confDir)).toBe(true);
        });

        it("does not create conf directory if in existence", () => {
            fs.mkdirSync(confDir);

            const mkdirSyncMock = jest.spyOn(fs, "mkdirSync");

            // eslint-disable-next-line no-new
            new Inventory(inventoryDir, confDir);

            expect(mkdirSyncMock).not.toHaveBeenCalled();
        });

        it("reads from inventory cache correctly", () => {

            const inventory = new Inventory(inventoryDir, confDir);

            expect(inventory.inventoryCacheFile).toEqual(inventoryConfFile);
        });
    });

    describe("modules", () => {
        let inventory: Inventory;

        beforeEach(() => {
            inventory = new Inventory(inventoryDir, confDir);

            if (fs.existsSync(inventoryConfFile)) {
                fs.unlinkSync(inventoryConfFile);
            }
        });

        it("shows modules", () => {
            expect(inventory.modules).toEqual([
                { name: "module-one" },
                { name: "module-two" }
            ]);
        });

        it("when no changes always shows the correct modules", () => {
            const modulesCallOne = inventory.modules;
            const modulesCallTwo = inventory.modules;

            expect(modulesCallOne).toEqual(modulesCallTwo);
        });

        it("invalidates cache when new module added", () => {
            const modulesCallOne = inventory.modules;

            const newModule = "module-three";
            addNewService(newModule, "service-five");

            inventory = new Inventory(inventoryDir, confDir);
            const modulesCallTwo = inventory.modules;

            expect(modulesCallOne).not.toEqual(modulesCallTwo);
            expect(modulesCallTwo).toContainEqual({ name: newModule });
        });
    });

    describe("services", () => {
        let inventory: Inventory;

        beforeEach(() => {
            inventory = new Inventory(inventoryDir, confDir);

            if (fs.existsSync(inventoryConfFile)) {
                fs.unlinkSync(inventoryConfFile);
            }
        });

        it("returns services with all dependencies", () => {

            const actual = inventory.services;

            for (const service of actual) {
                service.source = service.source.replace(inventoryDir, "inventory");
            }

            expect(actual).toMatchSnapshot();

        });

        it("invalidates cache when a new service is added", () => {

            const initialState = inventory.services;

            const newModule = "module-three";
            addNewService(newModule, "service-eight");

            inventory = new Inventory(inventoryDir, confDir);
            const amendedState = inventory.services;

            expect(amendedState).not.toEqual(initialState);
            const serviceEightExists = amendedState.some(service => service.name === "service-eight");
            expect(serviceEightExists).toBe(true);

        });

        it("invalidates cache when a service is modified", () => {
            const initialState = inventory.services;

            const initialService = initialState.find(service => service.name === "service-one");

            expect(initialService?.metadata.languageMajorVersion).toEqual(null);

            const servicePath = join(inventoryDir, "services/modules/module-one/service-one.docker-compose.yaml");

            const service = parse(fs.readFileSync(servicePath).toString("utf8"));

            service.services["service-one"].labels = [
                ...service.services["service-one"].labels,
                "chs.local.builder.languageVersion=17"
            ];

            fs.writeFileSync(
                servicePath,
                stringify(service)
            );

            const amendedState = inventory.services;

            expect(amendedState).not.toEqual(initialState);

            const amendedService = amendedState.find(service => service.name === "service-one");

            expect(amendedService?.metadata.languageMajorVersion).toEqual("17");
        });

        it("invalidates cache when service is removed", () => {
            const initialState = inventory.services;

            expect(initialState.find(service => service.name === "service-three")).toBeTruthy();

            const servicePath = join(inventoryDir, "services/modules/module-two/service-three.docker-compose.yaml");

            // @ts-expect-error does exist
            fs.rmSync(servicePath);

            inventory = new Inventory(inventoryDir, confDir);
            const amendedState = inventory.services;

            expect(amendedState).not.toEqual(initialState);

            expect(amendedState.find(service => service.name === "service-three")).toBeFalsy();
        });
    });
});
