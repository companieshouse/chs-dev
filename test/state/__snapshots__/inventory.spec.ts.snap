// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory services returns services with all dependencies 1`] = `
[
  {
    "builder": "",
    "dependsOn": [],
    "description": "primary database storing Companies House data",
    "metadata": {
      "healthcheck": "test $$(echo "rs.initiate({_id:'tilt-rs',members:[{_id:0,host:\\"mongo:27017\\"}]}).ok || rs.status().ok" | mongo --port 27017 --quiet) -eq 1",
      "ingressRoute": null,
      "languageMajorVersion": null,
      "repoContext": null,
    },
    "module": "infrastructure",
    "name": "mongo",
    "repository": null,
    "source": "inventory/services/infrastructure/mongo.docker-compose.yaml",
  },
  {
    "builder": "java",
    "dependsOn": [
      "mongo",
    ],
    "description": null,
    "metadata": {
      "healthcheck": "curl http://localhost:8080",
      "ingressRoute": "PathPrefix(\`/service-one/*\`)",
      "languageMajorVersion": null,
      "repoContext": null,
    },
    "module": "module-one",
    "name": "service-one",
    "repository": {
      "branch": null,
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-one/service-one.docker-compose.yaml",
  },
  {
    "builder": "node",
    "dependsOn": [
      "mongo",
      "service-one",
    ],
    "description": null,
    "metadata": {
      "healthcheck": null,
      "ingressRoute": null,
      "languageMajorVersion": "18",
      "repoContext": null,
    },
    "module": "module-one",
    "name": "service-two",
    "repository": {
      "branch": null,
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-one/service-two.docker-compose.yaml",
  },
  {
    "builder": "java",
    "dependsOn": [
      "mongo",
      "service-three",
      "service-two",
      "service-one",
    ],
    "description": null,
    "metadata": {
      "healthcheck": null,
      "ingressRoute": null,
      "languageMajorVersion": "17",
      "repoContext": null,
    },
    "module": "module-two",
    "name": "service-four",
    "repository": {
      "branch": null,
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-two/service-four.docker-compose.yaml",
  },
  {
    "builder": "",
    "dependsOn": [
      "mongo",
      "service-two",
      "service-one",
    ],
    "description": null,
    "metadata": {
      "healthcheck": null,
      "ingressRoute": null,
      "languageMajorVersion": null,
      "repoContext": "chs-dev/",
    },
    "module": "module-two",
    "name": "service-seven",
    "repository": {
      "branch": "feature/my-branch",
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-two/service-seven.docker-compose.yaml",
  },
  {
    "builder": "repository",
    "dependsOn": [
      "mongo",
      "service-one",
    ],
    "description": null,
    "metadata": {
      "healthcheck": null,
      "ingressRoute": null,
      "languageMajorVersion": null,
      "repoContext": null,
    },
    "module": "module-two",
    "name": "service-six",
    "repository": {
      "branch": null,
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-two/service-six.docker-compose.yaml",
  },
  {
    "builder": "java",
    "dependsOn": [
      "mongo",
      "service-two",
      "service-one",
    ],
    "description": null,
    "metadata": {
      "healthcheck": null,
      "ingressRoute": null,
      "languageMajorVersion": null,
      "repoContext": null,
    },
    "module": "module-two",
    "name": "service-three",
    "repository": {
      "branch": null,
      "url": "git@github.com:companieshouse/service-one.git",
    },
    "source": "inventory/services/modules/module-two/service-three.docker-compose.yaml",
  },
]
`;
